File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* a, ListNode* b) {
            return a->val > b->val;
        };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> min_heap(cmp);

        for (auto list : lists) {
            if (list != nullptr) {
                min_heap.push(list);
            }
        }
        auto dummy = new ListNode();
        auto curr = dummy;

        while (!min_heap.empty()) {
            auto min_node = min_heap.top();
            min_heap.pop();

            curr->next = min_node;
            curr = curr->next;

            if (min_node->next != nullptr) {
                min_heap.push(min_node->next);
            }
        }
        return dummy->next;
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
      # time complexity o(mlogk) m: the number of all eleemnt in resutl

        dummy = ListNode()
        cur = dummy
        min_heap = []

        for i, head in enumerate(lists):
            if head:
                heapq.heappush(min_heap, (head.val, i, head))

        while min_heap:
            val, idx, node = heapq.heappop(min_heap)

            cur.next = node
            cur = cur.next

            nxt_node = node.next
            if nxt_node:
                heapq.heappush(min_heap, (nxt_node.val, idx, nxt_node))
        
        return dummy.next

        